include/devicetree/pinctrl.h

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DEVICETREE_PINCTRL_H_
+#define ZEPHYR_INCLUDE_DEVICETREE_PINCTRL_H_
+
+/**
+ * @file
+ * @brief Devicetree pin control helpers
+ */
+
+/**
+ * @defgroup devicetree-pinctrl Pin control
+ * @ingroup devicetree
+ * @{
+ */
+
+/**
+ * @brief Get a node identifier for a phandle inside a pinctlr node by index
+ *
+ * Example devicetree fragment:
+ *
+ *     n: node {
+ *             pinctrl-0 = <&foo &bar>;
+ *             pinctrl-1 = <&baz &blub>;
+ *     }
+ *
+ * Example usage:
+ *
+ *     DT_PINCTRL_BY_IDX(DT_NODELABEL(n), 0, 1) // DT_NODELABEL(bar)
+ *     DT_PINCTRL_BY_IDX(DT_NODELABEL(n), 1, 0) // DT_NODELABEL(baz)
+ *
+ * @param node_id node with a pinctrl-'pc_idx' property
+ * @param pc_idx index of the pinctrl property itself
+ * @param idx index into the value of the pinctrl property
+ * @return node identifier for the phandle at index 'idx' in 'pinctrl-'pc_idx''
+ */
+#define DT_PINCTRL_BY_IDX(node_id, pc_idx, idx) \
+	DT_CAT6(node_id, _P_pinctrl_, pc_idx, _IDX_, idx, _PH)
+
+/**
+ * @brief Get a node identifier from a pinctrl-0 property
+ *
+ * This is equivalent to:
+ *
+ *     DT_PINCTRL_BY_IDX(node_id, 0, idx)
+ *
+ * It is provided for convenience since pinctrl-0 is commonly used.
+ *
+ * @param node_id node with a pinctrl-0 property
+ * @param idx index into the pinctrl-0 property
+ * @return node identifier for the phandle at index idx in the pinctrl-0
+ *         property of that node
+ */
+#define DT_PINCTRL_0_BY_IDX(node_id, idx) DT_PINCTRL_BY_IDX(node_id, 0, idx)
+
+/**
+ * @brief Get a node identifier from pinctrl-0 index 0
+ *
+ * This is equivalent to:
+ *
+ *     DT_PINCTRL_BY_IDX(node_id, 0, 0)
+ *
+ * It is provided for convenience since pinctrl-0 is commonly used.
+ *
+ * @param node_id node with a pinctrl-0 property
+ * @return node identifier for the phandle at index 0 in the pinctrl-0
+ *         property of that node
+ */
+#define DT_PINCTRL_0(node_id) DT_PINCTRL_BY_IDX(node_id, 0, 0)
```

Comment:
I'm not sure we should add this one. I think that `pinctrl` is better represented using states, so drivers should work using the state concept and not indices, so I'm not sure if providing extra helpers for indices helps on that front.

This code:
```
      n: node {
              pinctrl-0 = <&foo &bar>;
              pinctrl-1 = <&baz &blub>;
              pinctrl-names = "default", "sleep";
      }
 ```

and this other one:
```
      n: node {
              pinctrl-0 = <&foo &bar>;
              pinctrl-1 = <&baz &blub>;
              pinctrl-names = "sleep", "default";
      }
 ```

should result in same behavior, ie indices are not relevant.

Reply:
> so drivers should work using the state concept and not indices

This seems to be the main assertion here. The fact is that drivers are currently using indices, not just states. Now, that may be because there is no support for names yet. I don't know.

Does Linux use states exclusively, or does it also use indices?

Reply:
Yes, Linux uses states: https://www.kernel.org/doc/html/latest/driver-api/pin-control.html#pin-control-requests-from-drivers

As far as I know Zephyr has never had such concept (probably because DT API/tooling was missing and/or current pinctrl drivers are not in a mature status).

Reply:
I see this as no different than any other `name`d property in devicetree.  Its left to the binding and driver to decide if it wants to use the named version or not, and if the order of the values in an "array" has meaning.  For example regs and irqs do this all the time with assumptions in binding/driver.

Reply:
Whether it's the driver or not who takes care of storing states, names makes things easier. If every driver decides to use indices on their own, bindings will need to specify "use index 0 for default state and 1 for sleep".

If I have to write a driver that needs to obtain a particular IRQ (e.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/st/h7/stm32h7.dtsi#L452) I'll always obtain it by name, just to make sure I get what I want regardless of the order.

Reply:
> Whether it's the driver or not who takes care of storing states, names makes things easier. If every driver decides to use indices on their own, bindings will need to specify "use index 0 for default state and 1 for sleep".
> 
> If I have to write a driver that needs to obtain a particular IRQ (e.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/st/h7/stm32h7.dtsi#L452) I'll always obtain it by name, just to make sure I get what I want regardless of the order.

I agree that is better practice, but from an API point of view it makes sense to have what @mbolivar-nordic proposes in this PR.  How that gets used is left the the driver developers.

Reply:
> I agree that is better practice, but from an API point of view it makes sense to have what @mbolivar-nordic proposes in this PR. How that gets used is left the the driver developers.

If we think pinctrl representation is better understood using states, as in Linux, then indices should have no relevance. We now use index 0 everywhere because it's really the only thing we have, but as soon as more states start to be implemented then `0` will have no special meaning other than being the first entry. Summarizing, I think we should be careful on adding new APIs, specially because current pinctrl state is not mature enough. It may be useful as a transition step to make current code less verbose, though.

Reply:
Trying to summarize:

- @gmarull wants Zephyr to use states exclusively, thinks using indices is a bad practice, points at Linux pinmux APIs using states
- @galak agrees using names is better practice, but thinks this should be left up to the user to decide and the API should be neutral

I don't know what the project is going to decide re: states vs. indices.

However, if we do decide on a state-based approach, then doing all the index-based pin control access via dedicated APIs instead of directly accessing the pinctrl-0 and pinctrl-1 properties as we are doing now will make it trivial to deprecate the index-based way of doing things. So even if the API is not forever, I think it is better to introduce and use it than let the current status quo stand.

@gmarull would you agree with that argument?